cdap-app-fabric/src/main/java/io/cdap/cdap/app/job/SupportBundleJob.java

```diff
@@ -0,0 +1,471 @@
+        }
+      }
+    } else {
+      ProgramRunId latestProgramRunId =
+          new ProgramRunId(namespaceId, appId, ProgramType.WORKFLOW, workflowName, requestRunId);
+      runRecordDetailList = Arrays.asList(store.getRun(latestProgramRunId));
+    }
+
+    Queue<RunRecordDetail> runRecordDetailQueue = new LinkedList<>();
+    for (RunRecordDetail runRecordDetail : runRecordDetailList) {
+      runRecordDetailQueue.offer(runRecordDetail);
+    }
+    JsonObject subMetricCalculateTask = new JsonObject();
+    subMetricCalculateTask.addProperty("name", appId + "-metrics-collection");
+    subMetricCalculateTask.addProperty("type", "MetricCollectionTask");
+    subMetricCalculateTask.addProperty("startTimestamp", System.currentTimeMillis());
+    subMetricCalculateTask.addProperty("retries", 0);
+    subTasks.add(subMetricCalculateTask);
+
+    while (!runRecordDetailQueue.isEmpty()) {
+      RunRecordDetail runRecordDetail = runRecordDetailQueue.poll();
+      String runId = runRecordDetail.getProgramRunId().getRun();
+      Future<Pair<RunRecordDetail, JsonObject>> futureMetricsService =
+          executor.submit(
+              () -> {
+                JsonObject metrics = new JsonObject();
+                updateTask(subMetricCalculateTask, basePath, "IN_PROGRESS");
+                try {
+                  ApplicationDetail applicationDetail =
+                      applicationLifecycleService.getAppDetail(applicationId);
+                  metrics =
+                      supportBundlePipelineInfoTask.queryMetrics(
+                          metricsQueryHelper,
+                          namespaceId,
+                          appId,
+                          runId,
+                          workflowName,
+                          applicationDetail.getConfiguration(),
+                          runRecordDetail != null ? runRecordDetail.getStartTs() : 0,
+                          runRecordDetail != null && runRecordDetail.getStopTs() != null
+                              ? runRecordDetail.getStopTs()
+                              : DateTime.now().getMillis());
+                  retryServiceMap.remove(runId);
+                } catch (Exception e) {
+                  LOG.warn(
+                      String.format("Retried three times for this metrics with run id %s ", runId),
+                      e);
+                  queueTaskAfterFailed(runId, subMetricCalculateTask, basePath);
+                }
+                return new Pair<>(runRecordDetail, metrics);
+              });
+      if (!retryServiceMap.containsKey(runId)) {
+        futureMetricsList.add(futureMetricsService);
+      }
+    }
+    completeProcessing(futurePipelineInfoList, pipelineTaskMap, basePath, "pipelineinfo");
+    completeProcessingForMetrics(futureMetricsList, subMetricCalculateTask, basePath);
+  }
+
+  /** Generates run log file */
+  public void runLogProcessing(String basePath, String appPath) {
```

Comment:
Again, any task-specific code must be in task, job should do coordination and provide utility functions